### PR TITLE
Add support for casts in InPredicate to TupleDomain

### LINKED ISSUE #9979: Predicate pushdown does not work when column type is more specific than literal type
Consider below use case:
```
CREATE TABLE test (x bigint, col1 INTEGER, col4 SMALLINT);
SELECT * FROM test WHERE col4 IN (10002,10004);
```

Predicate from which `TupleDomain` is created is `(CAST("col4" AS integer) IN (10002, 10004))` and so Presto passes `TupleDomain.all()` when getting the table layout.

### PR DESCRIPTION
**Pull request by kokosing**
Add support for casts in InPredicate to TupleDomain

Fixes #9979

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 1]
```diff
@@ -685,10 +685,6 @@ private int compareOriginalValueToCoerced(Type originalValueType, Object origina
         @Override
         protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement)
         {
-            if (!(node.getValue() instanceof SymbolReference) || !(node.getValueList() instanceof InListExpression)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: most likely you should keep: `!(node.getValueList() instanceof InListExpression)`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 2]
```diff
@@ -1072,6 +1072,24 @@ public void testFromInPredicate()
 //        assertTrue(result.getTupleDomain().isNone());
     }
 
+    @Test
+    public void testFromInPredicateWithCasts()
+    {
+        Expression originalExpression = new InPredicate(
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: what about cases where there is `null` value in list (see commented code above), e.g:
```
A in (1L, 2L, null);
not (A in (1L, 2L, null);
A in (null);
not (A in (null));
```

**Comment by kokosing** [comment]
@sopel39 AC

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 3]
```diff
@@ -997,7 +998,7 @@ public void testFromUnprocessableInPredicate()
     {
         Expression originalExpression = new InPredicate(unprocessableExpression1(C_BIGINT), new InListExpression(ImmutableList.of(TRUE_LITERAL)));
         ExtractionResult result = fromPredicate(originalExpression);
-        assertEquals(result.getRemainingExpression(), originalExpression);
+        assertEquals(result.getRemainingExpression(), equal(unprocessableExpression1(C_BIGINT), TRUE_LITERAL));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: unrelated?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 4]
```diff
@@ -1049,27 +1050,68 @@ public void testFromInPredicate()
...
+        ExtractionResult result = fromPredicate(originalExpression);
+        assertEquals(result.getRemainingExpression(), TRUE_LITERAL);
+        assertEquals(result.getTupleDomain(), withColumnDomains(ImmutableMap.of(C_BIGINT, Domain.create(ValueSet.ofRanges(Range.equal(BIGINT, 1L), Range.equal(BIGINT, 2L)), false))));
+
+        originalExpression = not(in(C_BIGINT, Arrays.asList(1L, 2L, null)));
+        result = fromPredicate(originalExpression);
+        assertEquals(result.getRemainingExpression(), TRUE_LITERAL);
+        assertEquals(result.getTupleDomain(), TupleDomain.none());
+
+        originalExpression = isNull(in(C_BIGINT, Arrays.asList(1L, 2L, null)));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: wrapping in `isNull` doesn't make much sense as we ask for `C_BIGINT` domain. I wonder why it doesn't throw any error...

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 5]
```diff
@@ -1049,27 +1050,68 @@ public void testFromInPredicate()
...
+        assertEquals(result.getTupleDomain(), TupleDomain.all());
+
+        originalExpression = isNotNull(in(C_BIGINT, Arrays.asList(new Long[]{null})));
+        result = fromPredicate(originalExpression);
+        assertEquals(result.getRemainingExpression(), originalExpression);
+        assertEquals(result.getTupleDomain(), TupleDomain.all());
+    }
+
+    @Test
+    public void testFromInPredicateWithCasts()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: what about casts and null?

**Comment by kokosing** [comment]
@sopel39 Comments addressed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 6]
```diff
@@ -1114,6 +1114,38 @@ public void testFromInPredicateWithCasts()
         assertEquals(result.getTupleDomain(), withColumnDomains(ImmutableMap.of(C_SMALLINT, Domain.singleValue(SMALLINT, 1L))));
     }
 
+    @Test
+    public void testFromInPredicateWithCastsAndNulls()
+    {
+        Expression originalExpression = new InPredicate(
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: no... ; /

Cast insides makes it a bit harder.

**Comment by findepi** [comment]
i wasn't following and I am not fluent in `DomainTranslator`, so forgive stupid questions.

In case of expression `(CAST(someSmallintColumn AS integer) IN (10002, 10004))` (as in #9979), how would the `TupleDomain` look like?

In case of expression `(CAST(someBigintColumn AS integer) IN (10002, 10004))` (a narrowing cast), how would the `TupleDomain` look like? 

Also, today i realized that #9979  could have an alternative solution. @kokosing, you probably have thought about this, but noting  for completeness.
Alternatively (or supplementarily) we could change type of `10002` to be `SMALLINT`. In general, given integer constant, assign it a smallest type possible that can represent that constant (`TINYINT/SMALLINT/INTEGER/BIGINT`). Today we choose only `INTEGER/BIGINT`. That approach would remove the need for CAST altogether.

**Comment by kokosing** [comment]
> In case of expression (CAST(someSmallintColumn AS integer) IN (10002, 10004)) (as in #9979), how would the TupleDomain look like?

This not much different to test introduced here: https://github.com/prestodb/presto/pull/9984/commits/0d285b0ef9b28ae484a78eeabf088384c2724b56#diff-8df080055a9a25a18d663fa26716b197R870 And it is supporeted.

> In case of expression (CAST(someBigintColumn AS integer) IN (10002, 10004)) (a narrowing cast), how would the TupleDomain look like?

It is not supported.

**Comment by kokosing** [comment]
>  In general, given integer constant, assign it a smallest type possible that can represent that constant (TINYINT/SMALLINT/INTEGER/BIGINT). Today we choose only INTEGER/BIGINT. That approach would remove the need for CAST altogether.

Great idea. Anyway, some support for casting in `TupleDomain` and `DomainTranslator` is also needed.
